src/core/meshcop/joiner.cpp

```diff
@@ -388,74 +394,90 @@ void Joiner::HandleSecureCoapClientConnect(bool aConnected)
+    SuccessOrExit(error = mFinalizeMessage->Append(&vendorNameTlv, vendorNameTlv.GetSize()));
 
     vendorModelTlv.Init();
-    vendorModelTlv.SetVendorModel(mVendorModel);
-    SuccessOrExit(error = message->Append(&vendorModelTlv, vendorModelTlv.GetSize()));
+    vendorModelTlv.SetVendorModel(aVendorModel);
+    SuccessOrExit(error = mFinalizeMessage->Append(&vendorModelTlv, vendorModelTlv.GetSize()));
 
     vendorSwVersionTlv.Init();
-    vendorSwVersionTlv.SetVendorSwVersion(mVendorSwVersion);
-    SuccessOrExit(error = message->Append(&vendorSwVersionTlv, vendorSwVersionTlv.GetSize()));
+    vendorSwVersionTlv.SetVendorSwVersion(aVendorSwVersion);
+    SuccessOrExit(error = mFinalizeMessage->Append(&vendorSwVersionTlv, vendorSwVersionTlv.GetSize()));
 
     vendorStackVersionTlv.Init();
     vendorStackVersionTlv.SetOui(OPENTHREAD_CONFIG_STACK_VENDOR_OUI);
     vendorStackVersionTlv.SetMajor(OPENTHREAD_CONFIG_STACK_VERSION_MAJOR);
     vendorStackVersionTlv.SetMinor(OPENTHREAD_CONFIG_STACK_VERSION_MINOR);
     vendorStackVersionTlv.SetRevision(OPENTHREAD_CONFIG_STACK_VERSION_REV);
-    SuccessOrExit(error = message->Append(&vendorStackVersionTlv, vendorStackVersionTlv.GetSize()));
+    SuccessOrExit(error = mFinalizeMessage->Append(&vendorStackVersionTlv, vendorStackVersionTlv.GetSize()));
 
-    if (mVendorData != NULL)
+    if (aVendorData != NULL)
     {
         VendorDataTlv vendorDataTlv;
         vendorDataTlv.Init();
-        vendorDataTlv.SetVendorData(mVendorData);
-        SuccessOrExit(error = message->Append(&vendorDataTlv, vendorDataTlv.GetSize()));
+        vendorDataTlv.SetVendorData(aVendorData);
+        SuccessOrExit(error = mFinalizeMessage->Append(&vendorDataTlv, vendorDataTlv.GetSize()));
+    }
+
+    provisioningUrlTlv.Init();
+    provisioningUrlTlv.SetProvisioningUrl(aProvisioningUrl);
+
+    if (provisioningUrlTlv.GetLength() > 0)
+    {
+        SuccessOrExit(error = mFinalizeMessage->Append(&provisioningUrlTlv, provisioningUrlTlv.GetSize()));
     }
 
-    if (Get<Coap::CoapSecure>().GetDtls().mProvisioningUrl.GetLength() > 0)
+exit:
+    if (error != OT_ERROR_NONE && mFinalizeMessage != NULL)
     {
-        SuccessOrExit(error = message->Append(&Get<Coap::CoapSecure>().GetDtls().mProvisioningUrl,
-                                              Get<Coap::CoapSecure>().GetDtls().mProvisioningUrl.GetSize()));
+        mFinalizeMessage->Free();
     }
 
+    return error;
+}
+
+void Joiner::SendJoinerFinalize(void)
+{
+    VerifyOrExit(mFinalizeMessage != NULL);
+
 #if OPENTHREAD_ENABLE_CERT_LOG
-    LogCertMessage("[THCI] direction=send | type=JOIN_FIN.req |", *message);
+    LogCertMessage("[THCI] direction=send | type=JOIN_FIN.req |", *mFinalizeMessage);
 #endif
 
-    SuccessOrExit(error = Get<Coap::CoapSecure>().SendMessage(*message, Joiner::HandleJoinerFinalizeResponse, this));
+    SuccessOrExit(Get<Coap::CoapSecure>().SendMessage(*mFinalizeMessage, Joiner::HandleJoinerFinalizeResponse, this));
```

Comment:
Is it safe to reuse a sent message?

Reply:
I expect it should be. If `SendMessage()` returns an error it should not have taken ownership of the message. For example, in oroginal/previous implementation we would free the `message` if `SendMessage()` returned an error. 
The question is would `SendMessage()` perhaps change the message itself or does it ensure the message remains the same when/if it returns an error.

Reply:
Actully thinking about this, we should never get to situation to re-use the same `mFinalizeMessage` instance.

If `SendMessage()` is successful, we immediately set the `mFinalizeMessage` to `NULL`.

The case to cover is when `SendMessage()` returns an error. When we get to this situation and call `SendJoinerFinalize()` we are already at `OT_JOINER_STATE_CONNECTED` state and we have a timeout timer running. The only way forward is either being `Stop()`ed (which would clear/free the message) or timer timing out (which again free the message from `Finish()`).

So a `mFinalizeMessage` instance will only be used once.